File: MinStack/MinStack.py
class MinStack:

    # initialize your data structure here.
    def __init__(self):
        self.elements = []
        self.minIndexes = []

    # @param x, an integer
    # @return nothing
    def push(self, x):
        self.elements.append(x)
        if len(self.minIndexes) == 0 or x < self.elements[self.minIndexes[-1]]:
            self.minIndexes.append(len(self.elements) - 1)

    # @return nothing
    def pop(self):
        if self.minIndexes[-1] == len(self.elements) - 1:
            self.minIndexes.pop()
        self.elements.pop()

    # @return an integer
    def top(self):
        return self.elements[-1]

    # @return an integer
    def getMin(self):
        return self.elements[self.minIndexes[-1]]

File: MinStack/MinStack.rb
class MinStack
    # Initialize your data structure here
    def initialize
      @elements = []
      @minIndexes = []
    end

    # @param {Integer} x
    # @return {Void} nothing
    def push(x)
      @elements << x
      if @minIndexes.size == 0 || x < @elements[@minIndexes.last]
        @minIndexes << @elements.size - 1
      end
    end

    # @return {Void} nothing
    def pop
      if @minIndexes.last == @elements.size - 1
        @minIndexes.pop
      end
      @elements.pop
    end

    # @return {Integer}
    def top
      return @elements.last
    end

    # @return {Integer}
    def get_min
      return @elements[@minIndexes.last]
    end
end
